### PR TITLE
[PERF] base: avoid multiple ormcache computation in multithread

### PR DESCRIPTION
**Pull request by sts-odoo**
In multithreaded mode, when multiple queries arrive before the routing cache is filled,  multiple threads will try to compute the same methods.

This PR adds a single global lock on ormcache to avoid deadlock.

Comparing the requests on /shop on an empty database with website_sale,website_slides,website_event,website_hr_recruitment installed, the response time becomes higher with the number of concurrent requests. This table only compares value before the cache is used.

The tests were only invalidating the 'routing' cache

| concurrency | 1 | 2 | 3 | 4 | 5 |
|--------|--------|--------|--------|--------|--------|
| current situation - cached query only | 0.065117 | 0.0896905 | 0.131192 | 0.179595 | 0.247573
| current situation - non cached query only | 0.40558 | 0.733723 | 1.0675 | 1.40266 | 2.4364
| with the patch applied - cached query only | 0.0624025 | 0.087147 | 0.129508 | 0.18071 | 0.22474
| with the patch applied - non cached query only | 0.599373 | 0.652825 | 0.70965 | 0.766649 | 0.823916

image

image

For parallel workers, no significant difference could be observed with or without this patch.

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by rco-odoo** [comment]
Maybe you should make it clearer that the implementation adds one global lock per ormcached method.

**Comment by rco-odoo** [comment]
Two questions:
- Deadlocks are possible.  Do you want to prevent them?
- How can the response time of a _single_ request be faster with a lock ?  That doesn't make any sense.

File: odoo/tools/cache.py [hunk 1]
```diff
@@ -101,8 +102,12 @@ def lookup(self, method, *args, **kwargs):
...
         except KeyError:
             counter.miss += 1
-            value = d[key] = self.method(*args, **kwargs)
-            return value
+            #with args[0].env.registry._caches_lock:
+            with args[0].pool._Registry__caches_lock:  # to check with ci/security
+                with contextlib.suppress(KeyError):
+                    return d[key]
+                value = d[key] = self.method(*args, **kwargs)
+                return value
```
Comment (rco-odoo) [inline odoo/tools/cache.py:110]: Suggestion for the consistency of `counter`:
```suggestion
            with args[0].pool._Registry__caches_lock:  # to check with ci/security
                with contextlib.suppress(KeyError):
                    r = d[key]
                    counter.hit += 1
                    return r
                counter.miss += 1
                value = d[key] = self.method(*args, **kwargs)
                return value
```

**Review by rco-odoo — APPROVED** [review]
I agree with this version (modulo the little fix in the broken test).

File: odoo/tools/cache.py [hunk 2]
```diff
@@ -100,9 +101,14 @@ def lookup(self, method, *args, **kwargs):
             counter.hit += 1
             return r
         except KeyError:
-            counter.miss += 1
-            value = d[key] = self.method(*args, **kwargs)
-            return value
+            with args[0].pool._Registry__caches_lock:
```
Comment (HydrionBurst) [inline odoo/tools/cache.py:104]: Is it possible for a transaction has both `DROP TABLE` and ormcache usage?
Thread A holds the `_Registry__caches_lock` and waits for the `AccessShareLock` of the droped table
Thread B holds the `AccessExclusiveLock` for the dropped table and waits for the `_Registry__caches_lock`

**Pull request closed without merging**